src/platform/window.cpp

```diff
@@ -102,10 +111,76 @@ void Window::callbackKey(GLFWwindow *window, int key, int scancode, int action,
 
 void Window::callbackMousePosition(GLFWwindow *window, double xpos, double ypos) {
 	Window *w = reinterpret_cast<Window *>(glfwGetWindowUserPointer(window));
+
+	glm::vec2 absolute = glm::vec2(xpos, ypos);
+	glm::vec2 delta = absolute - w->_lastMousePosition.value_or(absolute);
+
+	if (w->_mousePositionCallback)
+		w->_mousePositionCallback(absolute, delta);
+
+	w->_lastMousePosition = absolute;
+}
+
+void Window::callbackMouseScroll(GLFWwindow *window, double xpos, double ypos) {
+	Window *w = reinterpret_cast<Window *>(glfwGetWindowUserPointer(window));
+	if (!w->_mouseScrollCallback)
+		return;
+
+	glm::vec2 position(xpos, ypos);
+	w->_mouseScrollCallback(position, position);
+}
+
+void Window::callbackMouseEnter(GLFWwindow *window, int entered) {
+	Window *w = reinterpret_cast<Window *>(glfwGetWindowUserPointer(window));
+	
+	if (entered) {
+		double x, y;
+		glfwGetCursorPos(w->_window, &x, &y);
+		w->_lastMousePosition = glm::vec2(x, y);
+	} else { // leaving the window
+		w->_lastMousePosition = {};
+	}
+
+	if (!w->_mouseEnterCallback)
+		return;
+
+	w->_mouseEnterCallback(entered);
 }
 
 void Window::callbackMouseButton(GLFWwindow *window, int button, int action, int mods) {
+	if (action == GLFW_REPEAT)
+		return;
+
+	Window *w = reinterpret_cast<Window *>(glfwGetWindowUserPointer(window));
+	if (!w->_mouseButtonCallback)
+		return;
+
+	w->_mouseButtonCallback(button, action, mods);
+}
+
+void Window::callbackGamepadButton(GLFWwindow *window, int button, int action) {
```

Comment:
I think you misunderstood me here, the gamepad callbacks should be moved into the GamepadMan and the Window should have absolutely nothing to do with the gamepads.